text/0000-else-match.md

```diff
@@ -0,0 +1,245 @@
+    }
+}
+```
+
+After:
+
+```rust
+if !self.typ.is_number() {
+    pieces.push(empty()); pieces.push(empty());
+} else match self.online {
+    Some(ref v) => {
+        pieces.push(v.mean().to_string());
+        pieces.push(v.stddev().to_string());
+    }
+    None => { pieces.push(empty()); pieces.push(empty()); }
+}
+```
+
+### [trust-dns](https://github.com/bluejekyll/trust-dns)
+
+<https://github.com/bluejekyll/trust-dns/blob/master/src/authority/authority.rs#L558>:
+
+Before:
+
+```rust
+if class == self.class {
+  match rr.get_rr_type() {
+    RecordType::ANY | RecordType::AXFR | RecordType::IXFR => return Err(ResponseCode::FormErr),
+    _ => (),
+  }
+} else {
+  match class {
+    DNSClass::ANY => {
+      if rr.get_ttl() != 0 { return Err(ResponseCode::FormErr) }
+      if let &RData::NULL(..) = rr.get_rdata() { () }
+      else { return Err(ResponseCode::FormErr) }
+      match rr.get_rr_type() {
+        RecordType::AXFR | RecordType::IXFR => return Err(ResponseCode::FormErr),
+        _ => (),
+      }
+    },
+    DNSClass::NONE => {
+      if rr.get_ttl() != 0 { return Err(ResponseCode::FormErr) }
+      match rr.get_rr_type() {
+        RecordType::ANY | RecordType::AXFR | RecordType::IXFR => return Err(ResponseCode::FormErr),
+        _ => (),
+      }
+    },
+    _ => return Err(ResponseCode::FormErr),
+  }
+}
+```
+
+After:
+
+```rust
+if class == self.class {
+  match rr.get_rr_type() {
+    RecordType::ANY | RecordType::AXFR | RecordType::IXFR => return Err(ResponseCode::FormErr),
+    _ => (),
+  }
+} else match class {
+  DNSClass::ANY => {
+    if rr.get_ttl() != 0 { return Err(ResponseCode::FormErr) }
+    if let &RData::NULL(..) = rr.get_rdata() { () }
+    else { return Err(ResponseCode::FormErr) }
+    match rr.get_rr_type() {
+      RecordType::AXFR | RecordType::IXFR => return Err(ResponseCode::FormErr),
+      _ => (),
+    }
+  },
+  DNSClass::NONE => {
+    if rr.get_ttl() != 0 { return Err(ResponseCode::FormErr) }
+    match rr.get_rr_type() {
+      RecordType::ANY | RecordType::AXFR | RecordType::IXFR => return Err(ResponseCode::FormErr),
+      _ => (),
+    }
+  },
+} else {
+  return Err(ResponseCode::FormErr);
+}
+```
+
+# Detailed design
+[design]: #detailed-design
+
+## Grammar
+
+See the following document for an (incomplete) guide to the grammar used in Rust:
+[Rust Documentation → Grammar](https://doc.rust-lang.org/grammar.html).
+
+This proposal modifies the
+[if expression grammar](https://doc.rust-lang.org/grammar.html#if-expressions).
+
+```
+else_tail : "else" [ if_expr | if_let_expr
++                  | match_expr
+                   | '{' block '}' ] ;
+```
+
+## Execution
+
+An `else match` block should be treated similar to an `else` block with a single `match`
+expression, with one key addition:
+
+When an arbitrary expression `exp` in `else match <exp>` fails to match any of the cases in the
+`else match` block, the next block (if it exists) is run.
```

Comment:
I am in favor of `else match`, but I am really against this! `match` should work consistently.

There is an orthogonal proposal to allow match to take `else`s at the end, so you don't have to include the catchall branch inside the match intend. But that proposal should apply to all `match` statements, not only the `else match` ones.